torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1152,45 +1174,46 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
 
   at::cuda::OptionalCUDAGuard gpuGuard;
 
-  pre(ncclStreams_[key]);
-
-  for (const auto i : c10::irange(inputs.size())) {
-    gpuGuard.set_index(devices[i].index());
-    at::cuda::CUDAStream& ncclStream = ncclStreams_[key][i];
-
-    // Both `inputs' and `outputs' are created on a worker stream and used in
-    // different ncclStreams.  Hence, both must record the ncclStream to
-    // prevent being freed before the collective finishes.
-    //
-    // We only record `inputs' here, and leave recording `outputs' to `fn' for
-    // operations where `inputs' and `outputs' are not the same.
-    //
-    // See [Sync Streams].
-    c10::cuda::CUDACachingAllocator::recordStream(
-        inputs[i].storage().data_ptr(), ncclStream);
-  }
+  pre(ncclStreams);
 
   {
     AutoNcclGroup nccl_group_guard;
     for (const auto i : c10::irange(inputs.size())) {
-      gpuGuard.set_index(devices[i].index());
-      at::cuda::CUDAStream& ncclStream = ncclStreams_[key][i];
-      C10D_NCCL_CHECK(
-          fn(inputs[i], outputs[i], ncclComms[i]->getNcclComm(), ncclStream), ncclComms[i]->getNcclCommFailureReason());
+      if (!inputs_same_dev || (inputs_same_dev && i == 0)) {
+        gpuGuard.set_index(devices[i].index());
+      }
+      decltype(i) stream_comm_i = (inputs_same_dev ? 0 : i);
```

Comment:
@mcarilli here is my noob question: is this the main change to take advantage of fusing multiple collective calls into single kernel?  if not, which change is the key part to take advantage of fusing?  

or users do not need to make any changes, if they queue a list of same type of collective calls (like all gather) in the same stream, NCCL will transparently fuse these calls into one kernel? 

Thanks!

Reply:
Never mind, just see document about NCCL group call https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/groups.html, so the fusing will work as long as we queue a list of same type of collective calls btw group start and group end